### PR TITLE
feat: DatePublication time zone support

### PR DESCRIPTION
**Pull request by killanaca**
Added support for GMT and BST label for the DatePublication component (_REPLAT-1155_)

> The United Kingdom is not on Greenwich Mean Time (GMT) all year. During Daylight Saving Time (DST) the correct time zone is British Summer Time BST (offset UTC+01).

The DatePublication accepts the date and time prop as a string in ISO 8601 (UTC), parse it and  apply London time zone.

**v2 reworked**: the DatePublication component displays time zone label **only if** the user have a different time zone than the current London time zone.

To get info about the user time zone, the component (try to) uses the method `Intl.DateTimeFormat().resolvedOptions().timeZone` and displays the time zone label as a fallback.

Note: When we parse the date **string** with `new Date()` or `Date.parse()` methods (they are equivalent), we get the date and time in the local time.

eg: on a browser with CST timezone (China Standard Time) 

```js
new Date("2017-01-01T12:00:00.000Z") // -> Sun Jan 01 2017 20:00:00 GMT+0800 (CST)
new Date(2017,0,1,12,0) // -> Sun Jan 01 2017 12:00:00 GMT+0800 (CST)
new Date("2017-01-01T12:00:00.000Z").getHours() // -> 20
new Date("2017-01-01T12:00:00.000Z").getUTCHours() // -> 12
new Date(2017,0,1,12,0).getHours() // -> 12
new Date(2017,0,1,12,0).getUTCHours() // -> 4
```

Fixed issues with different time zone in the running env (_REPLAT-655_)

bstgmt

In case of wrong date provided to the DatePublication component, this is the behaviour:
  
  
screen shot 2018-01-10 at 17 25 25

### CODE REVIEW

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://www.bitrise.io/build/0826c8f75bc009e6

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://www.bitrise.io/build/f674c3a8601456bf

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://www.bitrise.io/build/e6718b35bed30c97

File: packages/date-publication/__tests__/__snapshots__/date-publication.test.js.snap [hunk 1]
```diff
@@ -1,5 +1,13 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
-exports[`Date Publication test renders a DatePublication component with Sunday Times publication and Relevant date 1`] = `"Saturday July 01 2017, The Sunday Times"`;
+exports[`when the user have a different timezone than London timezone renders a DatePublication component with Times publication and Relevant GMT date 1`] = `"Sunday January 01 2017, 02:32pm GMT, The Times"`;
```
Comment (craigbilner) [inline packages/date-publication/__tests__/__snapshots__/date-publication.test.js.snap]: "when the user has...."

File: packages/date-publication/__tests__/date-helper.test.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/* eslint-env jest */
+
+import { isBST, getUTCTime } from "../date-helper";
```
Comment (craigbilner) [inline packages/date-publication/__tests__/date-helper.test.js]: minor: I'd imagine dates will be use in other places so should probably be in `utils`?

File: packages/date-publication/date-publication.js [hunk 3]
```diff
@@ -1,16 +1,42 @@
...
+    super(props);
+    this.state = {
+      tz: ""
+    };
+  }
+
+  componentDidMount() {
+    const dateUTC = getUTCTime(this.props.date);
+    if (!isLondonTimezone()) {
+      this.setState({ tz: isBST(dateUTC) ? " BST" : " GMT" }); // eslint-disable-line react/no-did-mount-set-state
```
Comment (craigbilner) [inline packages/date-publication/date-publication.js]: could put this in a function or wait for [another release](https://github.com/airbnb/javascript/issues/684)

File: packages/date-publication/date-publication.js [hunk 4]
```diff
@@ -1,16 +1,42 @@
...
+    }
+  }
+  render() {
+    const { date, publication } = this.props;
+    const datetimeUTC = getUTCTime(date);
+    const isDateGMT = !isBST(datetimeUTC);
```
Comment (craigbilner) [inline packages/date-publication/date-publication.js]: this is a little confusing. If the func is `isBST` think you should stay in that "mode" throughout the consuming code?

File: packages/date-publication/date-publication.js [hunk 5]
```diff
@@ -1,16 +1,42 @@
...
+    const offset = isDateGMT ? 0 : 60;
+    const datetimeLondonTimezone = addMinutes(datetimeUTC, offset);
```
Comment (craigbilner) [inline packages/date-publication/date-publication.js]: can we check the dates we hold in elastic search, do we really store UTC?

Comment (killanaca) [reply packages/date-publication/date-publication.js]: already checked at the beginning of the PR. Methode send us the date in UTC and according the content of the field in ES, we store the date (in ES) in UTC (the Z at the end)

File: packages/date-publication/__tests__/__snapshots__/date-publication.test.js.snap (hunk 1 above)
Comment (killanaca) [reply packages/date-publication/__tests__/__snapshots__/date-publication.test.js.snap]: thanks

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://www.bitrise.io/build/a96639cc5712f766

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://www.bitrise.io/build/c985a2f041e52e93

File: packages/date-publication/date-publication.js (hunk 3 above)
Comment (killanaca) [reply packages/date-publication/date-publication.js]: if we move the logic to a different function we get no-undef or react/sort-comp

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://www.bitrise.io/build/26ab2bce54692920

**Comment by times-tools** [comment]
Please find visual snapshots of your changed components here: https://www.bitrise.io/build/a7b04f4d28b1bf4e

**Pull request merged**